rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala

```diff
@@ -4,182 +4,229 @@ import coop.rchain.rholang.syntax.rholang_mercury.Absyn.{Ground => AbsynGround,
 
-    matcher.normalizeMatch(gs) should be (GString("String"))
+    GroundNormalizeMatcher.normalizeMatch(gs) should be(GString("String"))
   }
   "GroundUri" should "Compile as GUri" in {
     val gu = new GroundUri("Uri")
 
-    matcher.normalizeMatch(gu) should be (GUri("Uri"))
+    GroundNormalizeMatcher.normalizeMatch(gu) should be(GUri("Uri"))
   }
 }
 
-class ProcMatcherSpec extends FlatSpec with Matchers {
+class CollectMatcherSpec extends FlatSpec with Matchers {
   val inputs = ProcVisitInputs(
-      Par(),
-      DebruijnLevelMap[VarSort](),
-      DebruijnLevelMap[VarSort]())
+    Par(),
+    DebruijnLevelMap[VarSort]().newBindings(List((Some("P"), ProcSort), (Some("x"), NameSort)))._1,
+    DebruijnLevelMap[VarSort]())
+
+  "List" should "delegate" in {
+    val listData = new ListProc()
+    listData.add(new PVar("P"))
+    listData.add(new PEval(new NameVar("x")))
+    listData.add(new PGround(new GroundInt(7)))
+    val list = new PCollect(new CollectList(listData))
+
+    val result = ProcNormalizeMatcher.normalizeMatch(list, inputs)
+    result.par should be(
+      inputs.par.prepend(EList(List[Par](EVar(BoundVar(0)), Eval(ChanVar(BoundVar(1))), GInt(7)))))
+    result.knownFree should be(inputs.knownFree)
+  }
+
+  "Tuple" should "delegate" in {
+    val tupleData = new ListProc()
+    tupleData.add(new PVar("Q"))
+    tupleData.add(new PEval(new NameVar("y")))
+    val tuple = new PCollect(new CollectTuple(tupleData))
+
+    val result = ProcNormalizeMatcher.normalizeMatch(tuple, inputs)
+    result.par should be(
+      inputs.par.prepend(
+        ETuple(
+          List[Par](
+            EVar(FreeVar(0)),
+            Eval(ChanVar(FreeVar(1)))
+          ))))
+    result.knownFree should be(
+      inputs.knownFree.newBindings(List((Some("Q"), ProcSort), (Some("y"), NameSort)))._1, )
+  }
+  "Tuple" should "propgate free variables" in {
+    val tupleData = new ListProc()
+    tupleData.add(new PVar("Q"))
+    tupleData.add(new PGround(new GroundInt(7)))
+    tupleData.add(new PPar(new PGround(new GroundInt(7)), new PVar("Q")))
+    val tuple = new PCollect(new CollectTuple(tupleData))
+
+    an[Error] should be thrownBy {
+      ProcNormalizeMatcher.normalizeMatch(tuple, inputs)
+    }
+  }
+
+  "Set" should "delegate" in {
+    val setData = new ListProc()
+    setData.add(new PAdd(new PVar("P"), new PVar("R")))
+    setData.add(new PGround(new GroundInt(7)))
+    setData.add(new PPar(new PGround(new GroundInt(8)), new PVar("Q")))
+    val set = new PCollect(new CollectSet(setData))
+
+    val result = ProcNormalizeMatcher.normalizeMatch(set, inputs)
+    result.par should be(
+      inputs.par.prepend(
+        ESet(List[Par](EPlus(EVar(BoundVar(0)), EVar(FreeVar(0))),
+                       GInt(7),
+                       Par(GInt(8)).prepend((EVar(FreeVar(1))))))))
+    result.knownFree should be(
+      inputs.knownFree.newBindings(List((Some("R"), ProcSort), (Some("Q"), ProcSort)))._1, )
```

Comment:
I guess this one too - should be 

`result.knownFree should be (inputs.knownFree.newBindings(List((Some("R"), ProcSort), (Some("Q"), ProcSort)))._1)`

Reply:
Nvm this should go in a different PR.

Reply:
No, these were cleanups from the helpers, and I flubbed them.
I've fixed them now.